SymbolSolver is unable to .resolve()

Hello,

i am trying to find the fully qualified name of interface `main.MyInterface<E>` that is being implemented in class `main.MyClass<E>`.

```
package main;
public class MyClass<E> implements MyInterface<E> { }
```

```
package main;

public class MyInterface<E> { }
```

```
for (ClassOrInterfaceType type : classDeclaration.getImplementedTypes()) {
	ResolvedReferenceType rrt = type.resolve(); // exception gets thrown here
	if (rrt != null) {
		String interfaceName = rrt.getQualifiedName();
		// do other stuff with it
	}
}
```

Sadly, this yields me an exception:

```
Symbol resolution not configured: to configure consider setting a SymbolResolver in the ParserConfiguration
	at com.github.javaparser.ast.Node.lambda$getSymbolResolver$5(Node.java:682)
	at java.base/java.util.Optional.map(Optional.java:254)
	at com.github.javaparser.ast.Node.getSymbolResolver(Node.java:679)
	at com.github.javaparser.ast.type.ClassOrInterfaceType.resolve(ClassOrInterfaceType.java:292)
```

I am setting a SymbolSolver before I start running `.resolve()`:

```
CombinedTypeSolver typeSolver = new CombinedTypeSolver(new ReflectionTypeSolver());
// optionally some more external libraries added (required for what I am doing)
CombinedTypeSolver finalTypeSolver = new CombinedTypeSolver(typeSolver);
if (path.endsWith(".jar")) {
	finalTypeSolver.add(new JarTypeSolver(path));
} else if (path.endsWith("src")){
	finalTypeSolver.add(new JavaParserTypeSolver(path));
}

JavaSymbolSolver symbolSolver = new JavaSymbolSolver(finalTypeSolver);
JavaParser.getStaticConfiguration().setSymbolResolver(symbolSolver);
```

I also double checked without the generic types and it always yields the same exception. Same issue with extending from classes.
Am I doing something wrong or why does this not work as intended?


I just updated to the 3.7.1  (from 3.6.23) and now the exception is now this:

```
java.lang.IllegalStateException: No data of this type found. Use containsData to check for this first.
	at com.github.javaparser.ast.Node.getData(Node.java:457)
	at com.github.javaparser.ast.Node.lambda$getSymbolResolver$5(Node.java:699)
	at java.base/java.util.Optional.map(Optional.java:254)
	at com.github.javaparser.ast.Node.getSymbolResolver(Node.java:698)
	at com.github.javaparser.ast.type.ClassOrInterfaceType.resolve(ClassOrInterfaceType.java:292)
```


Since this did work before and only started to fail recently, I double checked what I did differently. And it turns out that I previously was using a path to a source folder (`JavaParserTypeSolver`). Now I am trying to read a JAR file and therefore use the `JarTypeSolver`, but it does not seem to work with that one. Whats going on here?

I have the same problem, except I am not using `SourceZip`. Below is the method that sets the symbol resolver:

```
private void setSymbolSolver() {
        final String sourceDirectory = getAnnotatedClassSourceInformation().getSourceDirectory();
        
        final ReflectionTypeSolver standardLibraryTypeSolver = new ReflectionTypeSolver();
        final JavaParserTypeSolver processedProjectTypeSolver = new JavaParserTypeSolver(sourceDirectory);

        final CombinedTypeSolver combinedTypeSolver = new CombinedTypeSolver(
            standardLibraryTypeSolver,
            processedProjectTypeSolver
        );

        final ClassLoader processedProjectClassLoader = getClass().getClassLoader();
        final URLClassLoader processedProjectURLClassLoader = (URLClassLoader) processedProjectClassLoader;

        final URL[] processedProjectDependencyLocations = processedProjectURLClassLoader.getURLs();

        try {
            for (final URL processedProjectDependencyLocation : processedProjectDependencyLocations) {
                final Path dependencyLocationPath = Paths.get(processedProjectDependencyLocation.toURI());

                if (dependencyLocationPath.endsWith(".jar")) {
                    final JavaParserTypeSolver dependencyTypeSolver = new JavaParserTypeSolver(dependencyLocationPath);

                    combinedTypeSolver.add(dependencyTypeSolver);
                }
            }
        }
        catch (final URISyntaxException e) {
            throw new RuntimeException(e);
        }

        final JavaSymbolSolver javaSymbolSolver = new JavaSymbolSolver(combinedTypeSolver);

        JavaParser.getStaticConfiguration().setSymbolResolver(javaSymbolSolver);
    }
```

and here is the calling code:

```
@Override
public Visitable visit(
    final ClassOrInterfaceType classOrInterfaceType,
    final PrimitiveTypesCombination primitiveTypesCombination
) {
    super.visit(classOrInterfaceType, primitiveTypesCombination);
    ResolvedReferenceType resolvedClassOrInterfaceType = classOrInterfaceType.resolve();
    ...
}
```

I get the following exception:
```
Caused by: java.lang.IllegalStateException: No data of this type found. Use containsData to check for this first.
```

--- a/javaparser-core-generators/src/main/java/com/github/javaparser/generator/core/visitor/CloneVisitorGenerator.java
+++ b/javaparser-core-generators/src/main/java/com/github/javaparser/generator/core/visitor/CloneVisitorGenerator.java
@@ -54,6 +54,7 @@ public class CloneVisitorGenerator extends VisitorGenerator {
 
         body.addStatement(builder.toString());
         body.addStatement("r.setComment(comment);");
+        body.addStatement("copyData(n, r);");
         body.addStatement("return r;");
     }
 }

--- a/javaparser-core-testing/src/test/java/com/github/javaparser/ast/DataKeyTest.java
+++ b/javaparser-core-testing/src/test/java/com/github/javaparser/ast/DataKeyTest.java
@@ -21,6 +21,7 @@
 
 package com.github.javaparser.ast;
 
+import com.github.javaparser.ast.expr.Name;
 import com.github.javaparser.ast.expr.SimpleName;
 import org.junit.jupiter.api.Test;
 
@@ -28,10 +29,13 @@ import java.util.Arrays;
 import java.util.List;
 
 import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;
+import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 
 class DataKeyTest {
     private static final DataKey<String> ABC = new DataKey<String>() {
+    };
+    private static final DataKey<String> DEF = new DataKey<String>() {
     };
     private static final DataKey<List<String>> LISTY = new DataKey<List<String>>() {
     };
@@ -60,4 +64,14 @@ class DataKeyTest {
         assertThrows(IllegalStateException.class, () -> node.getData(DING));
     }
 
+    @Test
+    void cloningCopiesData() {
+        Node node = new SimpleName();
+        node.setData(ABC, "ABC!");
+        node.setData(DEF, "DEF!");
+
+        Node clone = node.clone();
+        assertEquals("ABC!", clone.getData(ABC));
+        assertEquals("DEF!", clone.getData(DEF));
+    }
 }

--- a/javaparser-core-testing/src/test/java/com/github/javaparser/ast/visitor/CloneVisitorTest.java
+++ b/javaparser-core-testing/src/test/java/com/github/javaparser/ast/visitor/CloneVisitorTest.java
@@ -24,6 +24,7 @@ package com.github.javaparser.ast.visitor;
 import com.github.javaparser.ast.CompilationUnit;
 import com.github.javaparser.ast.NodeList;
 import com.github.javaparser.ast.body.*;
+import com.github.javaparser.ast.expr.Name;
 import com.github.javaparser.ast.type.Type;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;

--- a/javaparser-core/src/main/java/com/github/javaparser/ast/Node.java
+++ b/javaparser-core/src/main/java/com/github/javaparser/ast/Node.java
@@ -48,12 +48,13 @@ import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
 import static com.github.javaparser.ast.Node.Parsedness.PARSED;
 import static com.github.javaparser.ast.Node.TreeTraversal.PREORDER;
+import static java.util.Collections.emptySet;
 import static java.util.Collections.unmodifiableList;
 import static java.util.Spliterator.DISTINCT;
 import static java.util.Spliterator.NONNULL;
-import com.github.javaparser.ast.Node;
 import com.github.javaparser.metamodel.NodeMetaModel;
 import com.github.javaparser.metamodel.JavaParserMetaModel;
+import com.github.javaparser.ast.Node;
 
 /**
  * Base class for all nodes of the abstract syntax tree.
@@ -457,6 +458,18 @@ public abstract class Node implements Cloneable, HasParentNode<Node>, Visitable,
             throw new IllegalStateException("No data of this type found. Use containsData to check for this first.");
         }
         return value;
+    }
+
+    /**
+     * This method was added to support the clone method.
+     *
+     * @return all known data keys.
+     */
+    public Set<DataKey<?>> getDataKeys() {
+        if (data == null) {
+            return emptySet();
+        }
+        return data.keySet();
     }
 
     /**

--- a/javaparser-core/src/main/java/com/github/javaparser/ast/visitor/CloneVisitor.java
+++ b/javaparser-core/src/main/java/com/github/javaparser/ast/visitor/CloneVisitor.java
@@ -47,6 +47,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         CompilationUnit r = new CompilationUnit(n.getTokenRange().orElse(null), packageDeclaration, imports, types, module);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -58,6 +59,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         PackageDeclaration r = new PackageDeclaration(n.getTokenRange().orElse(null), annotations, name);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -69,6 +71,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         TypeParameter r = new TypeParameter(n.getTokenRange().orElse(null), name, typeBound, annotations);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -77,6 +80,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         LineComment r = new LineComment(n.getTokenRange().orElse(null), n.getContent());
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -85,6 +89,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         BlockComment r = new BlockComment(n.getTokenRange().orElse(null), n.getContent());
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -100,6 +105,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ClassOrInterfaceDeclaration r = new ClassOrInterfaceDeclaration(n.getTokenRange().orElse(null), modifiers, annotations, n.isInterface(), name, typeParameters, extendedTypes, implementedTypes, members);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -114,6 +120,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         EnumDeclaration r = new EnumDeclaration(n.getTokenRange().orElse(null), modifiers, annotations, name, implementedTypes, entries, members);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -126,6 +133,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         EnumConstantDeclaration r = new EnumConstantDeclaration(n.getTokenRange().orElse(null), annotations, name, arguments, classBody);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -138,6 +146,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         AnnotationDeclaration r = new AnnotationDeclaration(n.getTokenRange().orElse(null), modifiers, annotations, name, members);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -151,6 +160,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         AnnotationMemberDeclaration r = new AnnotationMemberDeclaration(n.getTokenRange().orElse(null), modifiers, annotations, type, name, defaultValue);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -162,6 +172,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         FieldDeclaration r = new FieldDeclaration(n.getTokenRange().orElse(null), modifiers, annotations, variables);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -173,6 +184,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         VariableDeclarator r = new VariableDeclarator(n.getTokenRange().orElse(null), type, name, initializer);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -189,6 +201,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ConstructorDeclaration r = new ConstructorDeclaration(n.getTokenRange().orElse(null), modifiers, annotations, typeParameters, name, parameters, thrownExceptions, body, receiverParameter);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -206,6 +219,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         MethodDeclaration r = new MethodDeclaration(n.getTokenRange().orElse(null), modifiers, annotations, typeParameters, type, name, parameters, thrownExceptions, body, receiverParameter);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -219,6 +233,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         Parameter r = new Parameter(n.getTokenRange().orElse(null), modifiers, annotations, type, n.isVarArgs(), varArgsAnnotations, name);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -229,6 +244,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         InitializerDeclaration r = new InitializerDeclaration(n.getTokenRange().orElse(null), n.isStatic(), body);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -237,6 +253,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         JavadocComment r = new JavadocComment(n.getTokenRange().orElse(null), n.getContent());
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -249,6 +266,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ClassOrInterfaceType r = new ClassOrInterfaceType(n.getTokenRange().orElse(null), scope, name, typeArguments, annotations);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -258,6 +276,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         PrimitiveType r = new PrimitiveType(n.getTokenRange().orElse(null), n.getType(), annotations);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -268,6 +287,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ArrayType r = new ArrayType(n.getTokenRange().orElse(null), componentType, n.getOrigin(), annotations);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -278,6 +298,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ArrayCreationLevel r = new ArrayCreationLevel(n.getTokenRange().orElse(null), dimension, annotations);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -288,6 +309,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         IntersectionType r = new IntersectionType(n.getTokenRange().orElse(null), elements);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -298,6 +320,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         UnionType r = new UnionType(n.getTokenRange().orElse(null), elements);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -307,6 +330,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         VoidType r = new VoidType(n.getTokenRange().orElse(null));
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -318,6 +342,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         WildcardType r = new WildcardType(n.getTokenRange().orElse(null), extendedType, superType, annotations);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -327,6 +352,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         UnknownType r = new UnknownType(n.getTokenRange().orElse(null));
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -337,6 +363,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ArrayAccessExpr r = new ArrayAccessExpr(n.getTokenRange().orElse(null), name, index);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -348,6 +375,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ArrayCreationExpr r = new ArrayCreationExpr(n.getTokenRange().orElse(null), elementType, levels, initializer);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -357,6 +385,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ArrayInitializerExpr r = new ArrayInitializerExpr(n.getTokenRange().orElse(null), values);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -367,6 +396,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         AssignExpr r = new AssignExpr(n.getTokenRange().orElse(null), target, value, n.getOperator());
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -377,6 +407,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         BinaryExpr r = new BinaryExpr(n.getTokenRange().orElse(null), left, right, n.getOperator());
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -387,6 +418,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         CastExpr r = new CastExpr(n.getTokenRange().orElse(null), type, expression);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -396,6 +428,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ClassExpr r = new ClassExpr(n.getTokenRange().orElse(null), type);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -407,6 +440,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ConditionalExpr r = new ConditionalExpr(n.getTokenRange().orElse(null), condition, thenExpr, elseExpr);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -416,6 +450,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         EnclosedExpr r = new EnclosedExpr(n.getTokenRange().orElse(null), inner);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -427,6 +462,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         FieldAccessExpr r = new FieldAccessExpr(n.getTokenRange().orElse(null), scope, typeArguments, name);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -437,6 +473,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         InstanceOfExpr r = new InstanceOfExpr(n.getTokenRange().orElse(null), expression, type);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -445,6 +482,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         StringLiteralExpr r = new StringLiteralExpr(n.getTokenRange().orElse(null), n.getValue());
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -453,6 +491,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         IntegerLiteralExpr r = new IntegerLiteralExpr(n.getTokenRange().orElse(null), n.getValue());
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -461,6 +500,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         LongLiteralExpr r = new LongLiteralExpr(n.getTokenRange().orElse(null), n.getValue());
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -469,6 +509,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         CharLiteralExpr r = new CharLiteralExpr(n.getTokenRange().orElse(null), n.getValue());
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -477,6 +518,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         DoubleLiteralExpr r = new DoubleLiteralExpr(n.getTokenRange().orElse(null), n.getValue());
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -485,6 +527,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         BooleanLiteralExpr r = new BooleanLiteralExpr(n.getTokenRange().orElse(null), n.getValue());
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -493,6 +536,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         NullLiteralExpr r = new NullLiteralExpr(n.getTokenRange().orElse(null));
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -505,6 +549,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         MethodCallExpr r = new MethodCallExpr(n.getTokenRange().orElse(null), scope, typeArguments, name, arguments);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -514,6 +559,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         NameExpr r = new NameExpr(n.getTokenRange().orElse(null), name);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -527,6 +573,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ObjectCreationExpr r = new ObjectCreationExpr(n.getTokenRange().orElse(null), scope, type, typeArguments, arguments, anonymousClassBody);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -536,6 +583,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         Name r = new Name(n.getTokenRange().orElse(null), qualifier, n.getIdentifier());
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -544,6 +592,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         SimpleName r = new SimpleName(n.getTokenRange().orElse(null), n.getIdentifier());
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -553,6 +602,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ThisExpr r = new ThisExpr(n.getTokenRange().orElse(null), classExpr);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -562,6 +612,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         SuperExpr r = new SuperExpr(n.getTokenRange().orElse(null), classExpr);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -571,6 +622,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         UnaryExpr r = new UnaryExpr(n.getTokenRange().orElse(null), expression, n.getOperator());
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -582,6 +634,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         VariableDeclarationExpr r = new VariableDeclarationExpr(n.getTokenRange().orElse(null), modifiers, annotations, variables);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -591,6 +644,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         MarkerAnnotationExpr r = new MarkerAnnotationExpr(n.getTokenRange().orElse(null), name);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -601,6 +655,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         SingleMemberAnnotationExpr r = new SingleMemberAnnotationExpr(n.getTokenRange().orElse(null), name, memberValue);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -611,6 +666,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         NormalAnnotationExpr r = new NormalAnnotationExpr(n.getTokenRange().orElse(null), name, pairs);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -621,6 +677,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         MemberValuePair r = new MemberValuePair(n.getTokenRange().orElse(null), name, value);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -632,6 +689,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ExplicitConstructorInvocationStmt r = new ExplicitConstructorInvocationStmt(n.getTokenRange().orElse(null), typeArguments, n.isThis(), expression, arguments);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -641,6 +699,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         LocalClassDeclarationStmt r = new LocalClassDeclarationStmt(n.getTokenRange().orElse(null), classDeclaration);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -651,6 +710,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         AssertStmt r = new AssertStmt(n.getTokenRange().orElse(null), check, message);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -660,6 +720,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         BlockStmt r = new BlockStmt(n.getTokenRange().orElse(null), statements);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -670,6 +731,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         LabeledStmt r = new LabeledStmt(n.getTokenRange().orElse(null), label, statement);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -678,6 +740,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         EmptyStmt r = new EmptyStmt(n.getTokenRange().orElse(null));
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -687,6 +750,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ExpressionStmt r = new ExpressionStmt(n.getTokenRange().orElse(null), expression);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -697,6 +761,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         SwitchStmt r = new SwitchStmt(n.getTokenRange().orElse(null), selector, entries);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -707,6 +772,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         SwitchEntry r = new SwitchEntry(n.getTokenRange().orElse(null), labels, n.getType(), statements);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -716,6 +782,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         BreakStmt r = new BreakStmt(n.getTokenRange().orElse(null), value);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -725,6 +792,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ReturnStmt r = new ReturnStmt(n.getTokenRange().orElse(null), expression);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -736,6 +804,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         IfStmt r = new IfStmt(n.getTokenRange().orElse(null), condition, thenStmt, elseStmt);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -746,6 +815,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         WhileStmt r = new WhileStmt(n.getTokenRange().orElse(null), condition, body);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -755,6 +825,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ContinueStmt r = new ContinueStmt(n.getTokenRange().orElse(null), label);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -765,6 +836,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         DoStmt r = new DoStmt(n.getTokenRange().orElse(null), body, condition);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -776,6 +848,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ForEachStmt r = new ForEachStmt(n.getTokenRange().orElse(null), variable, iterable, body);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -788,6 +861,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ForStmt r = new ForStmt(n.getTokenRange().orElse(null), initialization, compare, update, body);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -797,6 +871,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ThrowStmt r = new ThrowStmt(n.getTokenRange().orElse(null), expression);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -807,6 +882,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         SynchronizedStmt r = new SynchronizedStmt(n.getTokenRange().orElse(null), expression, body);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -819,6 +895,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         TryStmt r = new TryStmt(n.getTokenRange().orElse(null), resources, tryBlock, catchClauses, finallyBlock);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -829,6 +906,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         CatchClause r = new CatchClause(n.getTokenRange().orElse(null), parameter, body);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -839,6 +917,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         LambdaExpr r = new LambdaExpr(n.getTokenRange().orElse(null), parameters, body, n.isEnclosingParameters());
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -849,6 +928,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         MethodReferenceExpr r = new MethodReferenceExpr(n.getTokenRange().orElse(null), scope, typeArguments, n.getIdentifier());
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -858,6 +938,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         TypeExpr r = new TypeExpr(n.getTokenRange().orElse(null), type);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -879,6 +960,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ImportDeclaration r = new ImportDeclaration(n.getTokenRange().orElse(null), name, n.isStatic(), n.isAsterisk());
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -890,6 +972,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ModuleDeclaration r = new ModuleDeclaration(n.getTokenRange().orElse(null), annotations, name, n.isOpen(), directives);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -900,6 +983,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ModuleRequiresDirective r = new ModuleRequiresDirective(n.getTokenRange().orElse(null), modifiers, name);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -941,6 +1025,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ModuleExportsDirective r = new ModuleExportsDirective(n.getTokenRange().orElse(null), name, moduleNames);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -951,6 +1036,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ModuleProvidesDirective r = new ModuleProvidesDirective(n.getTokenRange().orElse(null), name, with);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -960,6 +1046,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ModuleUsesDirective r = new ModuleUsesDirective(n.getTokenRange().orElse(null), name);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -970,6 +1057,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ModuleOpensDirective r = new ModuleOpensDirective(n.getTokenRange().orElse(null), name, moduleNames);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -978,6 +1066,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         UnparsableStmt r = new UnparsableStmt(n.getTokenRange().orElse(null));
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -989,6 +1078,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         ReceiverParameter r = new ReceiverParameter(n.getTokenRange().orElse(null), annotations, type, name);
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -998,6 +1088,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         VarType r = new VarType(n.getTokenRange().orElse(null));
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -1006,6 +1097,7 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         Modifier r = new Modifier(n.getTokenRange().orElse(null), n.getKeyword());
         r.setComment(comment);
+        copyData(n, r);
         return r;
     }
 
@@ -1016,6 +1108,13 @@ public class CloneVisitor implements GenericVisitor<Visitable, Object> {
         Comment comment = cloneNode(n.getComment(), arg);
         SwitchExpr r = new SwitchExpr(n.getTokenRange().orElse(null), selector, entries);
         r.setComment(comment);
+        copyData(n, r);
         return r;
+    }
+
+    private void copyData(Node source, Node destination) {
+        for (DataKey dataKey : source.getDataKeys()) {
+            destination.setData(dataKey, source.getData(dataKey));
+        }
     }
 }

--- a/javaparser-core/src/main/java/com/github/javaparser/ast/visitor/ModifierVisitor.java
+++ b/javaparser-core/src/main/java/com/github/javaparser/ast/visitor/ModifierVisitor.java
@@ -31,11 +31,9 @@ import com.github.javaparser.ast.modules.*;
 import com.github.javaparser.ast.stmt.*;
 import com.github.javaparser.ast.type.*;
 import com.github.javaparser.utils.Pair;
-
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Optional;
-
 import static com.github.javaparser.utils.Utils.removeElementByObjectIdentity;
 import static com.github.javaparser.utils.Utils.replaceElementByObjectIdentity;
 

